server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -243,239 +231,82 @@ public void reconnectToNodes(DiscoveryNodes discoveryNodes, Runnable onCompletio
         });
     }
 
-    private enum ActivityType {
-        IDLE,
-        CONNECTING,
-        DISCONNECTING
-    }
-
-    /**
-     * {@link ConnectionTarget} ensures that we are never concurrently connecting to and disconnecting from a node, and that we eventually
-     * either connect to or disconnect from it according to whether {@link ConnectionTarget#connect(ActionListener)} or
-     * {@link ConnectionTarget#disconnect()} was called last.
-     * <p>
-     * Each {@link ConnectionTarget} is in one of these states:
-     * <p>
-     * - idle                       ({@link ConnectionTarget#future} has no listeners)
-     * - awaiting connection        ({@link ConnectionTarget#future} may contain listeners awaiting a connection)
-     * - awaiting disconnection     ({@link ConnectionTarget#future} may contain listeners awaiting a disconnection)
-     * <p>
-     * It will be awaiting connection (respectively disconnection) after calling {@code connect()} (respectively {@code disconnect()}). It
-     * will eventually become idle if these methods are not called infinitely often.
-     * <p>
-     * These methods return a {@link Runnable} which starts the connection/disconnection process iff it was idle before the method was
-     * called, and which notifies any failed listeners if the {@code ConnectionTarget} went from {@code CONNECTING} to {@code DISCONNECTING}
-     * or vice versa. The connection/disconnection process continues until all listeners have been removed, at which point it becomes idle
-     * again.
-     * <p>
-     * Additionally if the last step of the process was a disconnection then this target is removed from the current set of targets. Thus
-     * if this {@link ConnectionTarget} is idle and in the current set of targets then it should be connected.
-     * <p>
-     * All of the {@code listeners} are awaiting the completion of the same activity, which is either a connection or a disconnection.  If
-     * we are currently connecting and then {@link ConnectionTarget#disconnect()} is called then all connection listeners are
-     * removed from the list so they can be notified of failure; once the connecting process has finished a disconnection will be started.
-     * Similarly if we are currently disconnecting and then {@link ConnectionTarget#connect(ActionListener)} is called then all
-     * disconnection listeners are immediately removed for failure notification and a connection is started once the disconnection is
-     * complete.
-     */
```

Comment:
Treat the changes to `NodeConnectionsService` as a complete rewrite (with slightly different semantics, see test changes)